### PR TITLE
Index TokenBalances

### PR DESCRIPTION
**Pull request by igorffs**
#514 

## Motivation

This PR builds TokenBalances from TokenTransfer already existent data structure and creates in the database.

A TokenBalance is one of the 3 Addresses present in a TokenTransfer: `to_address_hash`, `from_address_hash` and `token_contract_address_hash`.  Every TokenTransfer is broken down and translated into this new entity.

In this PR only the records will be created in the database without their values (`value` and `value_fetched_at`), which will be fetched in the Realtime and Catchup indexers in a different PR, in order to unblock other issues that depend on this schema and improve the changes for the code review.

When the (already in progress #583) TokenBalanceFetcher is plugged to the indexers this infos will be fetched and updated.

### Enhancements

- We'll stop creating `Chain.Balance`s out of TokenTransfers, since TokenTransfer doesn't transfer Coin. If there's a Coin transfer in the same transaction where there is a TokenTransfer a Balance will be created from the Transaction.

### CODE REVIEW

File: apps/explorer/lib/explorer/chain/address/token_balance.ex [hunk 1]
```diff
@@ -0,0 +1,62 @@
+defmodule Explorer.Chain.Address.TokenBalance do
```
Comment (KronicDeth) [inline apps/explorer/lib/explorer/chain/address/token_balance.ex]: `Explorer.Chain.Balance` isn't scoped under `Explorer.Chain.Address` as `Explorer.Chain.Address.Balance`, so to keep the naming consistent I would make this `Explorer.Chain.TokenBalance`.    It is a `TokenBalance` at the pair of address and block number, so scoping only under address does not make sense to me.

File: apps/explorer/lib/explorer/chain/import.ex [hunk 2]
```diff
@@ -712,6 +744,29 @@ defmodule Explorer.Chain.Import do
...
+  @spec insert_token_balances([map()], %{
+          required(:timeout) => timeout(),
+          required(:timestamps) => timestamps()
+        }) ::
+          {:ok, [TokenBalance.t()]}
+          | {:error, [Changeset.t()]}
+  def insert_token_balances(changes_list, %{timeout: timeout, timestamps: timestamps})
+      when is_list(changes_list) do
+    # order so that row ShareLocks are grabbed in a consistent order
+    # ordered_changes_list = Enum.sort_by(changes_list, &{&1.transaction_hash, &1.log_index})
```
Comment (KronicDeth) [inline apps/explorer/lib/explorer/chain/import.ex]: You should always be ordering the changes list to prevent deadlocks.  This need to be uncommented and updated to use keys that make sense for token balances, like

```elixir
ordered_changes_lsit = Enum.sort_by(changes_list, &{&1.address_hash, &1.block_number})
```

File: apps/explorer/lib/explorer/chain/import.ex [hunk 3]
```diff
@@ -712,6 +744,29 @@ defmodule Explorer.Chain.Import do
...
+
+    {:ok, _} =
+      insert_changes_list(
+        changes_list,
```
Comment (KronicDeth) [inline apps/explorer/lib/explorer/chain/import.ex]: Use `ordered_changes_list` here.

File: apps/indexer/lib/indexer/block_fetcher/realtime.ex [hunk 4]
```diff
@@ -84,7 +85,8 @@ defmodule Indexer.BlockFetcher.Realtime do
            |> Map.drop(@import_options)
            |> put_in([:addresses, :params], balances_addresses_params)
            |> put_in([Access.key(:balances, %{}), :params], balances_params)
-           |> put_in([Access.key(:internal_transactions, %{}), :params], internal_transactions_params),
+           |> put_in([Access.key(:internal_transactions, %{}), :params], internal_transactions_params)
+           |> put_in([:token_balances, :params], token_balances_params),
```
Comment (KronicDeth) [inline apps/indexer/lib/indexer/block_fetcher/realtime.ex]: `:token_balances` `:params` is already in `options` at `line 65, so you don't need to `put_in` them again.  The pre-existing ones need to be `put_in` because they are updated by lines 69-82 of the `with`.  You're not updating the `token_balances_params`, so you don't need to `put_in` the value, it is already there.

File: apps/indexer/lib/indexer/block_fetcher.ex [hunk 5]
```diff
@@ -44,7 +44,8 @@ defmodule Indexer.BlockFetcher do
                 receipts: Import.receipts_options(),
                 token_transfers: Import.token_transfers_options(),
                 tokens: Import.tokens_options(),
-                transactions: Import.transactions_options()
+                transactions: Import.transactions_options(),
+                token_balances: Import.token_balances()
```
Comment (KronicDeth) [inline apps/indexer/lib/indexer/block_fetcher.ex]: Put `token_balances` before `token_transfers`, so it stay in the `token*` group and maintains alphabetical listing too.

File: apps/explorer/lib/explorer/chain/address/token_balance.ex [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+          address_hash: Hash.Address.t(),
+          token: %Ecto.Association.NotLoaded{} | Token.t(),
+          token_contract_address_hash: Hash.Address,
+          block_number: Block.block_number(),
+          inserted_at: DateTime.t(),
+          updated_at: DateTime.t(),
+          value: Wei.t() | nil
+        }
+
+  schema "address_token_balances" do
```
Comment (KronicDeth) [inline apps/explorer/lib/explorer/chain/address/token_balance.ex]: Since `TokenBalance` shouldn't be namespaces under `Address` neither should the table, rename this to `"token_balances"`.

File: apps/explorer/priv/repo/migrations/20180817021704_create_address_token_balances.exs [hunk 7]
```diff
@@ -0,0 +1,32 @@
+defmodule Explorer.Repo.Migrations.CreateAddressTokenBalances do
+  use Ecto.Migration
+
+  def change do
+    create table(:address_token_balances) do
```
Comment (KronicDeth) [inline apps/explorer/priv/repo/migrations/20180817021704_create_address_token_balances.exs]: Don't scope this under `:address_*`.  `:balances` isn't scoped under `:address_*`.

File: apps/indexer/lib/indexer/balances.ex [hunk 8]
```diff
@@ -40,9 +40,21 @@ defmodule Indexer.Balances do
                                                            is_binary(to_address_hash) and
                                                            is_binary(token_contract_address_hash) ->
       acc
-      |> MapSet.put(%{address_hash: from_address_hash, block_number: block_number})
-      |> MapSet.put(%{address_hash: to_address_hash, block_number: block_number})
-      |> MapSet.put(%{address_hash: token_contract_address_hash, block_number: block_number})
+      |> MapSet.put(%{
```
Comment (igorffs) [inline apps/indexer/lib/indexer/balances.ex]: Change this function test ☝️

File: apps/explorer/lib/explorer/chain/import.ex [hunk 9]
```diff
@@ -66,6 +67,12 @@ defmodule Explorer.Chain.Import do
           optional(:timeout) => timeout
         }
 
+  @type token_balances :: %{
+          required(:params) => params,
+          optional(:on_conflict) => :nothing | :replace_all,
```
Comment (alexgaribay) [inline apps/explorer/lib/explorer/chain/import.ex]: You can remove the `:on_conflict` since it's not being used.

**Comment by alexgaribay** [comment]
This may already be addressed in the follow-up PR but it seems like you need to keep track if a token transfer has already had its balances fetched for the respective hashes. Just a thought.

File: apps/explorer/lib/explorer/chain/address/token_balance.ex (hunk 1 above)
Comment (igorffs) [reply apps/explorer/lib/explorer/chain/address/token_balance.ex]: I brought this up in this PR to share a perspective about these columns naming and their relation to our domain knowledge and how it's represented in the code/database.

When I started tackling this issue I got a lead from @alexgaribay that this new requirement relates to the already existing Balance concept. I've started understanding the Balance from its schema which gave me a glance of was that about: 

> "The `Chain.Balance`  represents **an Address balance at a given point in time** (which is represented by its block_number).` It looks like a log!"

Reasoning about this and an explanation from @acravenho, telling me the difference and the kinds of values that we transfer are coin values (Ethereum, POA, and other coins) and token values (FREH tokens), gave me the resources I needed to understand what we were building now and how both concepts are related.

> "The `Chain.TokenBalance`  represents an **Address Token' balance at a given point in time** (which is represented by its block_number).` Oh, it's also a log!"

My first thought of an approach for dealing this was straightforward:
-  We already have a `Chain.Balance`
- Now we'll have a `Chain.TokenBalance`

Then I noticed two things:
1. Since now we have a TokenBalance, Balance gets too generic.
2. Both concepts have a strong relation to Address and are actually its content (we could even interpret them as addresses table attributes). But that knowledge isn't explicit to the developers through the architecture.

When I look at this is now, this is how it looks:

screen shot 2018-08-23 at 10 51 31


My proposal is: Let's communicate that knowledge through the architecture making these relations more explicit in the code organization and also the database by nesting Balance in the Address instead of making the TokenBalance generic. Also changing Balance name to CoinBalance so both concepts differences get clearer.

Then, we'll be able to look at these aspects this way:

screen shot 2018-08-23 at 10 58 33

What do you folks think? cc @KronicDeth @feliperenan @acravenho @alexgaribay

Comment (KronicDeth) [reply apps/explorer/lib/explorer/chain/address/token_balance.ex]: I'm fine with also renaming the old Balances, I didn't want the naming to be inconsistent and I assumed you wouldn't want to change the Balance code in this PR since Plataformatec has considered those sorta changes in PRs I've done as putting too much in one PR, but renaming the old Balances to be more specific also works and was something I would have proposed.

I do prefer moving common suffixes or prefixes to be namespaces though, so I'd go with

* `Explorer.Chain.Address.Balance.Coin` - `address_balance_coins`
* `Explorer.Chain.Address.Balance.Token` - `address_balance_tokens`

Potential con: You do have a limited length for identifier of 63 bytes in Postgres, so namespacing balances under address could lead to index/constraint naming problems.  Since we don't have other types of balances (once you include coin and token in the name) besides those associated with `{address, block}` pairs, not having the `address_` is 8 characters we could save.

Comment (acravenho) [reply apps/explorer/lib/explorer/chain/address/token_balance.ex]: I don't think this should be scoped under `Explorer.Chain.Address` as this does not correctly describe the relationship between `block`, `address`, and `balances`.

`Explorer.Chain.CoinBalance` and `Explorer.Chain.TokenBalance` seems like it would make sense here for the reasons @igorffs outlined above.

Comment (acravenho) [reply apps/explorer/lib/explorer/chain/address/token_balance.ex]: Oops, I should have refreshed the page before submitting.

Comment (KronicDeth) [reply apps/explorer/lib/explorer/chain/address/token_balance.ex]: I agree with @acravenho that scoping under `Address` loses that  the block is also important.  

Since we can't scope 2 dimensionally under both `Block` and `Address`, but the `Coin` rename is a really good idea because a top-level `Balance` is confusing, we could go with

* `Explorer.Chain.Balance.Coin` - `balance_coins`
* `Explorer.Chain.Balance.Token` - balance_tokens`

... but there's the big elephant of the Token scope.

If we make `Coin` one scope and `Token` another we get

* `Explorer`
  * `Chain`
    * `Coin`
      * `Balance`
    * `Token`
      * `Balance`
      * `Transfer`

This tree emphasizes the differences between the base `Coin` and `Token`s while using `Address` or `Balance` as a seems to be more cross cutting.

I would have pushed for a `Token` namespace eventually anyway as it follows the rule of using a namespace when you start to repeat prefixes or suffixes.

**Comment by igorffs** [comment]
> This may already be addressed in the follow-up PR but it seems like you need to keep track if a token transfer has already had its balances fetched for the respective hashes. Just a thought.

@alexgaribay I'm not sure if it is and if we need it I believe we'll have to make sure that this happens in this PR because it's the one that is responsible to stores in the database (the other one is just getting the value in the blockchain). By that you mean we just have to make sure it stored before the TokenTransfer or is there more to it?

But I'm not sure if we need to worry about it. 🤔 

I like to think that the relation between the TokenBalance and the TokenTransfer in our indexing context is that the transfer is an event that triggers a necessity to update a value that we want to show to our users. The informational value the TokenBalance doesn't depend whether the  TokenTransfer is indexed or not because what makes it relevant is that we show the current balance values in the user's wallet. It's very similar of the way we have in the Address page today, where we get the value directly from the blockchain. 

Do you think this :point_up: makes sense?

**Comment by alexgaribay** [comment]
@igorffs Yes. I looked more closely at the code and the sibling branch. Everything looks good 👍 .

File: apps/explorer/lib/explorer/chain/address/token_balance.ex (hunk 1 above)
Comment (igorffs) [reply apps/explorer/lib/explorer/chain/address/token_balance.ex]: When I get myself trying to understand where to put "something" I try to question what kind of responsibility it has or what kind of information it provides me. I do that by trying to find the dimension that provides me a more "complete"  informational value in the smallest possible unity (a single record in this scenario). Let's take a TokenBalance for instance:

We have the following attributes: `address_hash`, `value`, `value_fetched_at`, `block_number`. Let's look at the possible dimensions, considering we want to get  a value:

- Asking "what does it mean?" to `address` + `value` tells me what is the address token balance. When we add other relations to this, we give more meaning to this information:
  - By adding the `block_number` we add a "when/where" meaning and we'll be able to identify when (in which block) that was the address' token value.
  - By adding a `token_contract` we give a "what kind of value this is" meaning and we'll be able to identify what token was used so we could have that balance. 
- Asking "what does it mean" to `block_number` + `value` doesn't tell me what is the block token balance. We need to add an "address" attribute to give more meaning to it.
  - The same logic is applied to the token_contract + value. :point_up:

That may confuse you a little, but there are a few other shreds of evidence that made me think like that:

- If we didn't need this info to behave like a log, we would probably add this in the Address table having: `Address.token_balances_values`  the same way we have an `Address.balance_value`  today.
- To get this info from the blockchain using `eth_getBalance` or the `balanceOf` of a token contract, we need to pass the resource identification, which is the address hash. We even make a single RPC call for each address.
- The context where we primarily want to show this info is in the Address page.

In summary, all relations are indeed very important. But the strongest relation here is the Address, that's why I believe that it needs to be scoped under Address.

Does that change your perception? @acravenho  @KronicDeth 🤔

Comment (KronicDeth) [reply apps/explorer/lib/explorer/chain/address/token_balance.ex]: `Address.fetched_balance` only exists because I didn't want to delete the field when I made the balances table: it was for compatibility with a legacy decision, not because I thought it was still necessary once we started to care about historic balances.

You're making the argument that because `Address` is the most important, scope it to `Address`.  I will give you that, but I am (and I think @acravenho from https://github.com/poanetwork/blockscout/pull/582#discussion_r212346571) prefers that when there is multi-dimensional scoping to not favor one of the associations even when one is stronger and instead leave the schema at unnested under `Chain`.

The problem I have with arguing the supremacy of Address means nest something under Address is that the argument can be made for other schemas - a Block is mined by an Address, a Transaction is from an Address and only that `from` can sign the transaction, the other addresses don't sign off on it.

I defer to whatever @acravenho wants because I don't think I can convince you non-nesting due to multi-dimensional scoping is a better argument than nesting due to supremacy.  I think we just have different styles for namespacing.

Comment (igorffs) [reply apps/explorer/lib/explorer/chain/address/token_balance.ex]: > The problem I have with arguing the supremacy of Address means nest something under Address is that the argument can be made for other schemas - a Block is mined by an Address, a Transaction is from an Address and only that from can sign the transaction, the other addresses don't sign off on it.

@KronicDeth what I'm trying to present is a perspective to model a how we make a dependency explicit in our architecture given how the problem is presented to ourselves. This is different from the Block and Transaction because they're both strong entities themselves where the Address has a role in these contexts (an Address in the Block is a Miner and in the transaction can be one participant), not an attribute related to that single Address. 

> I defer to whatever @acravenho wants because I don't think I can convince you non-nesting due to multi-dimensional scoping is a better argument than nesting due to supremacy. I think we just have different styles for namespacing.

I tend to model technical solutions as close as possible the knowledge earned from the domain we're dealing. I believe that makes it easier for developers and non-technical stakeholders to build the solution together. That has a lot of inspiration in DDD and ubiquitous language.

But in order to that to work out we need to be on the same page and if that doesn't represent yall perception of an ideal architecture I'd be ok in keeping the proposed mindset.  💪

File: apps/explorer/lib/explorer/chain/address/token_balance.ex [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+  """
+
+  use Ecto.Schema
+  import Ecto.Changeset
+
+  alias Explorer.Chain.Address.TokenBalance
+  alias Explorer.Chain.{Address, Block, Hash, Token, Wei}
+
+  @typedoc """
+   *  `address` - The `t:Explorer.Chain.Address.t/0` that is the balance's owner.
```
Comment (feliperenan) [inline apps/explorer/lib/explorer/chain/address/token_balance.ex]: Should we have a primary key here? I'm not sure how attribute would fit if it, but it could be a simple `id` 🤔 

Wdyt? @igorffs and @KronicDeth

File: apps/explorer/lib/explorer/chain/address/token_balance.ex (hunk 1 above)
Comment (igorffs) [inline apps/explorer/lib/explorer/chain/address/token_balance.ex]: @KronicDeth regarding the concern of making other changes that don't necessarily relate to this PR objective: I'll open a different PR right after to make the required adjustments when we get in a conclusion in this other discussion ☝️

File: apps/explorer/lib/explorer/chain/address/token_balance.ex (hunk 10 above)
Comment (KronicDeth) [reply apps/explorer/lib/explorer/chain/address/token_balance.ex]: We don't have a primary key for `Balance`s.  If you add it here, you need to add it there too.

Since it hasn't been necessary for `Balance`, it's probably not necessary here either.   Nothing in the code is likely to need an ID because you'll look up `Address.Balance.Token` by `address_hash` the same as `Balance` is now.

File: apps/explorer/lib/explorer/chain/address/token_balance.ex (hunk 1 above)
Comment (KronicDeth) [reply apps/explorer/lib/explorer/chain/address/token_balance.ex]: I am fine with you making the namespace changes in this PR.  I got the impression your team had issues with refactoring mixed with other changes.  I'm fine with renaming happening in the same PR, I would only leave it in separate commits.

**Comment by KronicDeth** [comment]
Fixed the missed `mix format` in a rebase.  We'll see if Mainnet and Sokol will cooperate on the timeouts this time still.

**Comment by KronicDeth** [comment]
Disabled flaky Geth test (opened https://github.com/poanetwork/blockscout/issues/597) and testing again.
